applications/usb_uart_bridge/src/main.c

```diff
@@ -0,0 +1,260 @@
+					     UART_BUF_SIZE - (*rx)->len);
+		(*rx)->len += data_length;
+
+		if ((*rx)->len > 0) {
+			if (((*rx)->len == UART_BUF_SIZE) ||
+			   ((*rx)->buffer[(*rx)->len - 1] == '\n') ||
+			   ((*rx)->buffer[(*rx)->len - 1] == '\r') ||
+			   ((*rx)->buffer[(*rx)->len - 1] == '\0')) {
+				k_fifo_put(peer_dev_data->fifo, (*rx));
+				k_sem_give(&peer_dev_data->sem);
+
+				(*rx) = NULL;
+			}
+		}
+	}
+
+	if (uart_irq_tx_ready(dev)) {
+		struct uart_data *buf = k_fifo_get(dev_data->fifo, K_NO_WAIT);
+		u16_t written = 0;
+
+		/* Nothing in the FIFO, nothing to send */
+		if (!buf) {
+			uart_irq_tx_disable(dev);
+			return;
+		}
+
+		while (buf->len > written) {
+			written += uart_fifo_fill(dev,
+						  &buf->buffer[written],
+						  buf->len - written);
+		}
+
+		while (!uart_irq_tx_complete(dev)) {
+			/* Wait for the last byte to get
+			 * shifted out of the module
+			 */
+		}
+
+		if (k_fifo_is_empty(dev_data->fifo)) {
+			uart_irq_tx_disable(dev);
+		}
+
+		k_free(buf);
+	}
+}
+
+void power_thread(void)
+{
+	while (1) {
+		u32_t status = NRF_POWER->USBREGSTATUS;
+
+		if (0 == (status & POWER_USBREGSTATUS_VBUSDETECT_Msk)) {
+			NRF_POWER->SYSTEMOFF = 1;
+		}
+		k_sleep(100);
+	}
+}
+
+void main(void)
+{
+	int ret;
+	struct serial_dev *usb_0_dev_data = &devs[0];
+	struct serial_dev *usb_1_dev_data = &devs[1];
+	struct serial_dev *uart_0_dev_data = &devs[2];
+	struct serial_dev *uart_1_dev_data = &devs[3];
+	struct device *usb_0_dev, *usb_1_dev, *uart_0_dev, *uart_1_dev;
+
+	usb_0_dev = device_get_binding("CDC_ACM_0");
+	if (!usb_0_dev) {
+		printk("CDC ACM device not found\n");
+		return;
+	}
+
+	usb_1_dev = device_get_binding("CDC_ACM_1");
+	if (!usb_1_dev) {
+		printk("CDC ACM device not found\n");
+		return;
+	}
+
+	uart_0_dev = device_get_binding("UART_0");
+	if (!uart_0_dev) {
+		printk("UART 0 init failed\n");
+	}
+
+	uart_1_dev = device_get_binding("UART_1");
+	if (!uart_1_dev) {
+		printk("UART 1 init failed\n");
+	}
+
+	usb_0_dev_data->dev = usb_0_dev;
+	usb_0_dev_data->fifo = &usb_0_tx_fifo;
+	usb_0_dev_data->peer = uart_0_dev_data;
+
+	usb_1_dev_data->dev = usb_1_dev;
+	usb_1_dev_data->fifo = &usb_1_tx_fifo;
+	usb_1_dev_data->peer = uart_1_dev_data;
+
+	uart_0_dev_data->dev = uart_0_dev;
+	uart_0_dev_data->fifo = &uart_0_tx_fifo;
+	uart_0_dev_data->peer = usb_0_dev_data;
+
+	uart_1_dev_data->dev = uart_1_dev;
+	uart_1_dev_data->fifo = &uart_1_tx_fifo;
+	uart_1_dev_data->peer = usb_1_dev_data;
+
+	k_sem_init(&usb_0_dev_data->sem, 0, 1);
+	k_sem_init(&usb_1_dev_data->sem, 0, 1);
+	k_sem_init(&uart_0_dev_data->sem, 0, 1);
+	k_sem_init(&uart_1_dev_data->sem, 0, 1);
+
+	/* For unknown reasons, we have to sleep ~200 ms here
+	 * for the USB to be recognized properly by Windows. We loose early
+	 * log information on power cycle because of this.
+	 * TODO: Investiate and fix this bug.
+	 */
+	k_sleep(200);
+
+	/* Wait 1 sec for the host to do all settings */
+	k_busy_wait(1000000);
```

Comment:
Why not `k_sleep` here in stead? You can use the `K_SECONDS` macro here for clarity.

Reply:
Good points. I will try to clean this up and retest it

Reply:
Wait statements proved unnecessary after some testing. Probably something USB-related has been fixed upstream in the meantime.